### ISSUE TITLE
Dataset workflow plugins

### ISSUE THREAD

**Issue by wardi** [ckan/ideas-and-roadmap#108]
I'm proposing a new plugin interface to define a site's dataset editing and publishing workflows. In my case "workflow" means:
- one or more approval stages on new datasets and edits to datasets within an organization
- last approval stage might be a timed release "publish at date/time: X"
- arrival at each approval stage may notify users that are responsible for that stage
- users in the organization can tell what stage an approval is waiting on
- users in the organization may view the the currently published and pending versions

These workflows could be configured per organization/dataset type with plugins that register which types or orgs they handle. This is something like the way resource view plugins work with resource types.

On the dataset editing form when changes are pending the plugins enabled would choose which version to present to the user (possibly with highlighting of fields that have changed) and expose new buttons besides "save" "delete", such as "publish", "submit for approval", "save for later", "suggest deletion". The default "save"/"delete" would likely be disabled for normal users. Users should be able to switch back to the "real" current version of the dataset to see the original values or possibly submit different competing changes into the approval process.

On the dataset view page users in the organization would be able to select the pending changes to view. Plugins would determine the default for a given user.

On the dataset search page the changes submitted would be indexed like private datasets. We would need urls for suggested changes so that we can link directly to them, e.g. `/dataset/traffic-data@suggested993'
## Changeset data model

Most workflow plugins could be implemented on top of a single new model I've called "changeset" (feel free to suggest a better name)

| Changeset field | notes |
| --- | --- |
| user_id | user that submitted this change |
| name | slug to identify the change (e.g. "suggested993") |
| created_date | date |
| target_id | id of the dataset being modified (perhaps orgs, groups in future) |
| target_type | `'dataset'` |
| plugin_type | identifier to find plugin responsible for managing this changeset |
| jsonpatch | changes to make ([in a format that works for edits and deletion](https://github.com/ckan/ckan/pull/1679#issuecomment-47279807)?) |
| approvals | list of approval strings (postgres text array?) |
| scheduled_date | for future scheduled publish/delete events, null normally |
| plugin_data | extra data (json) as required by plugin |
| deleted | bool |

Changesets can be active (able/unable to apply) or deleted (withdrawn/rejected). Unable to apply means that the jsonpatch 'test' conditions fail for the current version of the dataset.

Changesets are _not_ hierarchical - they describe changes that apply directly to the current version of the dataset. It may be possible for a plugin to implement changesets that are displayed in a hierarchical manner using information in plugin_data or in other tables, but supporting such use is not part of the design of this feature.
## IWorkflow methods

(to be completed)
## Example: suggested changes

This kind of workflow plugin would allow the creation of up to one changeset per user per dataset: "User A suggests the following changes to dataset B"

Users that have permission to view a dataset but not edit it would be presented with a "suggest changes" button on the dataset view page. Clicking would lead to the normal dataset and resource edit pages, but instead of "save" user would have a "suggest change" button.  After suggesting a change the edit screen for this dataset would also have a "withdraw suggested change" button.

When a user has a suggested change when viewing that dataset they may toggle between their suggested version or the currently published version of that dataset.

Users with permission to edit a dataset may view all users' suggested change versions. When viewing suggested changes they may "reject change" or "apply change".

Notifications: suggestions would be displayed on the dashboards of users with permissions to change the dataset. If a suggestion is accepted or rejected the change author could be notified by email. If a suggested change no longer applies (the original fields modified no longer match the values being changed) the change author could also be notified by email.
## Example: moderated edits (single approval)

This kind of workflow plugin would replicate the "moderated edits" workflow: one "pending" version of a dataset exists that users in the same organization may view and edit. Admins may approve pending versions.

Users in the organization viewing a dataset can switch between the one official pending version of the dataset and the current published version. They may suggest deletion of a dataset, create a new shared pending version of a dataset (automatically deleting the existing one) or edit/withdraw an existing pending version.

Users in the organization may create a blank, private dataset that will exist to reserve a dataset id and allow creation of a to-be-published version based on a changeset.  

Admins would be able to apply pending changesets on the dataset edit page or with a bulk approval tool.
## Example: multiple approval

Similar to single approval above, but approvals are stored in the changeset until a changeset receives enough approvals (a determined by the workflow plugin) to actually publish the change.

Users in the organization would be able to see what approvals the pending changeset has received so far.

The workflow plugin would decide who to notify as as changeset is approved.
## Example: scheduled approval

In any workflow plugin instead of immediately applying a change, a changeset may be marked with a `scheduled_date`. There will be a common mechanism to automatically attempt to apply any changeset with a scheduled_date in the past.

A workflow plugin supporting scheduled changes would have to decide whether to support edits on a new changeset while a scheduled change is still pending, or even to support multiple scheduled changes on the same dataset. A simpler approach may be to abort the scheduled change or disallow new changes until the scheduled change has been published.

**Comment by JJIguiniz** [comment]
Would it be possible to get a time stamp for each approval step or change in order to get basic performance measures out of the workflow? That way you could in theory see how long each step takes, which users, or organizations are more efficient, and use the plugin to inform six sigma style continuous improvement.

**Comment by wardi** [comment]
We aren't doing any work on this for the moment. This is a tricky thing to generalize and it might be better to have smaller, more focused features to address specific parts of this proposal.

**Comment by rbvictor** [comment]
Hello there! Could there be any chance of this issue being addressed in the short term? 
If not, is it possible to implement these features by creating extensions or must the core be modified in order to accomplish this?

**Comment by wardi** [comment]
alternate approach in #211 that is simpler and allows for lots of other great features at the same time.

**Comment by romeosanjose** [comment]
this is similar of what we have implemented and created plugin (notification system included) for a government agency site way back in november 2015.
